lxd/auth/bearer/bearer.go

```diff
@@ -0,0 +1,108 @@
+package bearer
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+
+	"github.com/canonical/lxd/lxd/identity"
+	"github.com/canonical/lxd/lxd/request"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// IsRequest returns true if the caller sent a bearer token in the Authorization header that is a JWT and appears to
+// have this LXD cluster as the issuer. If true, it returns the raw token, and the subject.
+func IsRequest(r *http.Request, clusterUUID string) (isRequest bool, token string, subject string) {
+	// Check Authorization header for bearer token.
+	token, ok := strings.CutPrefix(r.Header.Get("Authorization"), "Bearer ")
+	if !ok || token == "" {
+		return false, "", ""
+	}
+
+	// Check we can parse the token. If we can't parse it, it could be an opaque OAuth2.0 token.
+	claims := jwt.MapClaims{}
+	t, _, err := jwt.NewParser().ParseUnverified(token, claims)
+	if err != nil {
+		return false, "", ""
+	}
+
+	// There must be an issuer
+	issuer, err := t.Claims.GetIssuer()
+	if err != nil {
+		return false, "", ""
+	}
+
+	// There must be a subject
+	sub, err := t.Claims.GetSubject()
+	if err != nil {
+		return false, "", ""
+	}
+
+	// Expect the issuer to be "lxd:{cluster_uuid}".
+	expectIssuer := "lxd:" + clusterUUID
+	if issuer != expectIssuer {
+		return false, "", ""
+	}
+
+	// Expect the audience to be "lxd:{cluster_uuid}" (the token is signed symmetrically).
+	gotAudience, err := t.Claims.GetAudience()
+	if err != nil {
+		return false, "", ""
+	}
+
+	if len(gotAudience) != 1 || gotAudience[0] != expectIssuer {
+		return false, "", ""
+	}
+
+	return true, token, sub
+}
+
+// Authenticate gets a bearer identity from the cache using the given subject, and verifies that it is of the expected
+// type. It then verifies that the token was signed by the secret associated with that identity, and that the token has
+// not expired.
+func Authenticate(token string, subject string, identityCache *identity.Cache, expectedIdentityType string) (*request.RequestorArgs, error) {
+	// Get the identity from the cache by the subject.
+	entry, err := identityCache.Get(api.AuthenticationMethodBearer, subject)
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate that it is of the expected type. (When/if we have token authentication to the main API, we want to make
```

Comment:
@markylaing wrt to this point and the `"lxd:{cluster_uuid}"` audience string above.

How about we use a different audience for devlxd tokens vs future general API tokens?

We could then validate the tokens are for use with the specific API earlier in the request.

E.g. `"devlxd:{cluster_uuid}"`

Reply:
That is a very good point and is exactly what the audience is for!! Wonderful. Will update. Thanks.